### DISCUSSION TITLE
Anonymous types should implement the interface they are assigned to

### DISCUSSION THREAD

**Discussion by AArnott** [dotnet/csharplang#1542] [Language Ideas]
It's almost like duck-typing, but still strongly typed. When I create an anonymous type that the C# compiler can see must implement a particular interface in order to compile, the generated anonymous type should implement that interface implicitly.

For example, this should compile:

```cs
public interface IDiffBaseInfo
{
    string Id { get; set; }
    string Description { get; set; }
}

public IDiffBaseInfo GetCommit()
{
    return new { Id = "a", Description = "b" };
}
```

**Comment by HaloFour** [comment]
Anonymous types in C# are readonly.

I could possibly see this working if the properties were declared as readonly also, but I'd prefer a more explicit syntax for creating the anonymous type, e.g. `new IDiffBaseInfo { ... }` like Java anonymous classes.

**Comment by bondsbw** [comment]
Similar: #100

**Comment by AArnott** [comment]
#100 seems to be dealing with omitting the type in a `new` expression where the type is defined elsewhere. It doesn't deal with anonymous types.

**Comment by Joe4evr** [comment]
[dotnet/roslyn#13 (comment)](https://github.com/dotnet/roslyn/issues/13#issuecomment-173413281)

**Comment by AArnott** [comment]
Ah ya, that looks like a superset of what I'm asking for here. I'm certainly not asking to implement interfaces with *methods* on them. But the response to it doesn't bode well for this smaller) request.

**Comment by bondsbw** [comment]
@AArnott Right, but #100 gives the caller a similar syntax to work with.  It is still more work on the definition side.

**Comment by glucaci** [comment]
It's still something happening with this? @AArnott @gafter 

In #100 you should provide an implementation to can use the `new` keyword in that way.

I find this construct `new IFoo { Bar = "Foo" }` very useful for tests.

**Comment by jnm2** [comment]
@glucaci You are asking for https://github.com/dotnet/roslyn/issues/13#issuecomment-173413281:

> We have no expectation of ever doing anything like this.

**Comment by glucaci** [comment]
@jnm2 yes, thanks for pointing that.
I saw that the dotnet/roslyn#13 feature request was closed without any sustainable explanation.

**Comment by glucaci** [comment]
Following @MadsTorgersen https://github.com/dotnet/roslyn/issues/13#issuecomment-70338359

> You could certainly imagine an evolution of anonymous objects in C# that would combine the current behavior with the ability to inherit classes, implement interfaces and supply member implementations.

maybe we can have a more recent update about what he is thinking on this topic.

Pointing to local classes (https://github.com/dotnet/roslyn/issues/13#issuecomment-144432884) has nothing to do with this request because than is useless, I can declare a class anywhere else and use it.
The idea is to not implement that interface at all and still be able to construct objects.

**Comment by gafter** [comment]
> maybe we can have a more recent update about what he is thinking on this topic.

I suspect it is unlikely he has thought about it since then.
